File: packages/core/test/test-bundle.js
import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import { encodeBundle, decodeBundle, validateBundle } from '../src/bundle.js'
import { BundleError } from '../src/errors.js'
import { BUNDLE_MAGIC, BUNDLE_VERSION, HEADER_SIZE } from '@wlearn/types'

describe('encodeBundle + decodeBundle', () => {
  it('round-trips manifest and artifacts', () => {
    const manifest = { typeId: 'wlearn.test@1', params: { lr: 0.1 } }
    const artifacts = [
      { id: 'model', data: new Uint8Array([1, 2, 3, 4]) },
      { id: 'config', data: new Uint8Array([10, 20]), mediaType: 'application/json' }
    ]

    const bytes = encodeBundle(manifest, artifacts)
    const { manifest: m, toc, blobs } = decodeBundle(bytes)

    assert.equal(m.typeId, 'wlearn.test@1')
    assert.equal(m.bundleVersion, BUNDLE_VERSION)
    assert.deepEqual(m.params, { lr: 0.1 })
    assert.equal(toc.length, 2)

    // Artifacts sorted by id: config before model
    assert.equal(toc[0].id, 'config')
    assert.equal(toc[1].id, 'model')

    const configBlob = blobs.subarray(toc[0].offset, toc[0].offset + toc[0].length)
    assert.deepEqual([...configBlob], [10, 20])
    assert.equal(toc[0].mediaType, 'application/json')

    const modelBlob = blobs.subarray(toc[1].offset, toc[1].offset + toc[1].length)
    assert.deepEqual([...modelBlob], [1, 2, 3, 4])
  })

  it('handles empty artifacts', () => {
    const manifest = { typeId: 'wlearn.empty@1' }
    const bytes = encodeBundle(manifest, [])
    const { manifest: m, toc } = decodeBundle(bytes)
    assert.equal(m.typeId, 'wlearn.empty@1')
    assert.equal(toc.length, 0)
  })

  it('preserves header magic and version', () => {
    const bytes = encodeBundle({ typeId: 'wlearn.test@1' }, [])
    assert.equal(bytes[0], 0x57) // W
    assert.equal(bytes[1], 0x4c) // L
    assert.equal(bytes[2], 0x52) // R
    assert.equal(bytes[3], 0x4e) // N

    const view = new DataView(bytes.buffer)
    assert.equal(view.getUint32(4, true), BUNDLE_VERSION)
  })

  it('produces deterministic output', () => {
    const manifest = { typeId: 'wlearn.det@1', params: { b: 2, a: 1 } }
    const artifacts = [
      { id: 'z', data: new Uint8Array([1]) },
      { id: 'a', data: new Uint8Array([2]) }
    ]
    const a = encodeBundle(manifest, artifacts)
    const b = encodeBundle(manifest, artifacts)
    assert.deepEqual(a, b)
  })
})

describe('validateBundle', () => {
  it('passes on valid bundle', () => {
    const manifest = { typeId: 'wlearn.test@1' }
    const artifacts = [{ id: 'data', data: new Uint8Array([42, 43, 44]) }]
    const bytes = encodeBundle(manifest, artifacts)
    const { manifest: m, toc } = validateBundle(bytes)
    assert.equal(m.typeId, 'wlearn.test@1')
    assert.equal(toc.length, 1)
  })

  it('throws on corrupted blob', () => {
    const bytes = encodeBundle(
      { typeId: 'wlearn.test@1' },
      [{ id: 'data', data: new Uint8Array([1, 2, 3]) }]
    )
    // Corrupt the last byte (inside blob region)
    bytes[bytes.length - 1] ^= 0xff
    assert.throws(() => validateBundle(bytes), BundleError)
  })
})

describe('decodeBundle validation', () => {
  it('rejects truncated header', () => {
    assert.throws(() => decodeBundle(new Uint8Array(10)), BundleError)
  })

  it('rejects bad magic', () => {
    const bytes = encodeBundle({ typeId: 'wlearn.test@1' }, [])
    bytes[0] = 0x00
    assert.throws(() => decodeBundle(bytes), BundleError)
  })

  it('rejects bad version', () => {
    const bytes = encodeBundle({ typeId: 'wlearn.test@1' }, [])
    const view = new DataView(bytes.buffer)
    view.setUint32(4, 99, true)
    assert.throws(() => decodeBundle(bytes), (err) => {
      assert(err instanceof BundleError)
      assert(err.message.includes('99'))
      return true
    })
  })

  it('rejects when manifestLen + tocLen exceeds bytes', () => {
    const bytes = encodeBundle({ typeId: 'wlearn.test@1' }, [])
    const view = new DataView(bytes.buffer)
    view.setUint32(8, 999999, true) // set manifestLen way too large
    assert.throws(() => decodeBundle(bytes), BundleError)
  })

  it('rejects overlapping TOC entries', () => {
    // Build a valid bundle then manually craft overlapping TOC
    // This is hard to trigger via encodeBundle, so we construct manually
    const manifest = JSON.stringify({ typeId: 'wlearn.test@1', bundleVersion: 1 })
    const toc = JSON.stringify([
      { id: 'a', offset: 0, length: 10, sha256: 'abc' },
      { id: 'b', offset: 5, length: 10, sha256: 'def' }
    ])
    const enc = new TextEncoder()
    const mBytes = enc.encode(manifest)
    const tBytes = enc.encode(toc)
    const blob = new Uint8Array(20)

    const total = HEADER_SIZE + mBytes.length + tBytes.length + blob.length
    const buf = new Uint8Array(total)
    const view = new DataView(buf.buffer)
    buf.set(BUNDLE_MAGIC, 0)
    view.setUint32(4, 1, true)
    view.setUint32(8, mBytes.length, true)
    view.setUint32(12, tBytes.length, true)
    buf.set(mBytes, HEADER_SIZE)
    buf.set(tBytes, HEADER_SIZE + mBytes.length)
    buf.set(blob, HEADER_SIZE + mBytes.length + tBytes.length)

    assert.throws(() => decodeBundle(buf), BundleError)
  })

  it('rejects TOC entry out of bounds', () => {
    const manifest = JSON.stringify({ typeId: 'wlearn.test@1', bundleVersion: 1 })
    const toc = JSON.stringify([
      { id: 'a', offset: 0, length: 100, sha256: 'abc' }
    ])
    const enc = new TextEncoder()
    const mBytes = enc.encode(manifest)
    const tBytes = enc.encode(toc)
    const blob = new Uint8Array(5) // only 5 bytes, but TOC says 100

    const total = HEADER_SIZE + mBytes.length + tBytes.length + blob.length
    const buf = new Uint8Array(total)
    const view = new DataView(buf.buffer)
    buf.set(BUNDLE_MAGIC, 0)
    view.setUint32(4, 1, true)
    view.setUint32(8, mBytes.length, true)
    view.setUint32(12, tBytes.length, true)
    buf.set(mBytes, HEADER_SIZE)
    buf.set(tBytes, HEADER_SIZE + mBytes.length)
    buf.set(blob, HEADER_SIZE + mBytes.length + tBytes.length)

    assert.throws(() => decodeBundle(buf), BundleError)
  })
})

File: py/tests/test_bundle.py
import hashlib
import struct

import pytest

from wlearn.bundle import encode_bundle, decode_bundle, validate_bundle, HEADER_SIZE
from wlearn.errors import BundleError


def test_round_trip():
    manifest = {'typeId': 'test.model@1', 'params': {'C': 1.0}}
    artifacts = [
        {'id': 'weights', 'data': b'\x01\x02\x03'},
        {'id': 'bias', 'data': b'\x04\x05'},
    ]
    bundle = encode_bundle(manifest, artifacts)
    m, toc, blobs = decode_bundle(bundle)

    assert m['typeId'] == 'test.model@1'
    assert m['bundleVersion'] == 1
    assert m['params'] == {'C': 1.0}
    assert len(toc) == 2
    # artifacts sorted by id: bias before weights
    assert toc[0]['id'] == 'bias'
    assert toc[1]['id'] == 'weights'
    assert bytes(blobs[toc[0]['offset']:toc[0]['offset'] + toc[0]['length']]) == b'\x04\x05'
    assert bytes(blobs[toc[1]['offset']:toc[1]['offset'] + toc[1]['length']]) == b'\x01\x02\x03'


def test_empty_artifacts():
    manifest = {'typeId': 'test.empty@1'}
    bundle = encode_bundle(manifest, [])
    m, toc, blobs = decode_bundle(bundle)
    assert m['typeId'] == 'test.empty@1'
    assert len(toc) == 0


def test_header_magic_and_version():
    manifest = {'typeId': 'test.hdr@1'}
    bundle = encode_bundle(manifest, [{'id': 'x', 'data': b'\x00'}])
    assert bundle[:4] == b'WLRN'
    version = struct.unpack_from('<I', bundle, 4)[0]
    assert version == 1


def test_determinism():
    manifest = {'typeId': 'test.det@1', 'b': 2, 'a': 1}
    arts = [{'id': 'z', 'data': b'\x01'}, {'id': 'a', 'data': b'\x02'}]
    b1 = encode_bundle(manifest, arts)
    b2 = encode_bundle(manifest, arts)
    assert b1 == b2


def test_validate_passes():
    manifest = {'typeId': 'test.val@1'}
    arts = [{'id': 'model', 'data': b'hello world'}]
    bundle = encode_bundle(manifest, arts)
    m, toc, blobs = validate_bundle(bundle)
    assert m['typeId'] == 'test.val@1'
    expected_hash = hashlib.sha256(b'hello world').hexdigest()
    assert toc[0]['sha256'] == expected_hash


def test_validate_corrupted_blob():
    manifest = {'typeId': 'test.corrupt@1'}
    arts = [{'id': 'model', 'data': b'hello world'}]
    bundle = bytearray(encode_bundle(manifest, arts))
    # corrupt last byte of blob
    bundle[-1] ^= 0xFF
    with pytest.raises(BundleError, match='SHA-256 mismatch'):
        validate_bundle(bytes(bundle))


def test_reject_truncated_header():
    with pytest.raises(BundleError, match='too small'):
        decode_bundle(b'WLR')


def test_reject_bad_magic():
    buf = bytearray(HEADER_SIZE)
    buf[:4] = b'NOPE'
    with pytest.raises(BundleError, match='Invalid bundle magic'):
        decode_bundle(bytes(buf))


def test_reject_bad_version():
    buf = bytearray(HEADER_SIZE)
    buf[:4] = b'WLRN'
    struct.pack_into('<I', buf, 4, 99)
    struct.pack_into('<I', buf, 8, 0)
    struct.pack_into('<I', buf, 12, 0)
    with pytest.raises(BundleError, match='Unsupported bundle version'):
        decode_bundle(bytes(buf))


def test_reject_truncated_manifest():
    buf = bytearray(HEADER_SIZE)
    buf[:4] = b'WLRN'
    struct.pack_into('<I', buf, 4, 1)
    struct.pack_into('<I', buf, 8, 9999)  # manifest way too long
    struct.pack_into('<I', buf, 12, 0)
    with pytest.raises(BundleError, match='truncated'):
        decode_bundle(bytes(buf))


def test_reject_overlapping_toc():
    """Build a bundle with manually overlapping TOC entries."""
    import json
    manifest_json = json.dumps(
        {'typeId': 'test.overlap@1', 'bundleVersion': 1},
        sort_keys=True, separators=(',', ':')).encode()
    toc_json = json.dumps([
        {'id': 'a', 'offset': 0, 'length': 10, 'sha256': 'x'},
        {'id': 'b', 'offset': 5, 'length': 10, 'sha256': 'y'},
    ], sort_keys=True, separators=(',', ':')).encode()

    header = struct.pack('<4sIII', b'WLRN', 1, len(manifest_json), len(toc_json))
    blob_data = b'\x00' * 20
    bundle = header + manifest_json + toc_json + blob_data

    with pytest.raises(BundleError, match='overlap'):
        decode_bundle(bundle)


def test_reject_toc_out_of_bounds():
    """Build a bundle with a TOC entry pointing past the blob region."""
    import json
    manifest_json = json.dumps(
        {'typeId': 'test.oob@1', 'bundleVersion': 1},
        sort_keys=True, separators=(',', ':')).encode()
    toc_json = json.dumps([
        {'id': 'a', 'offset': 0, 'length': 9999, 'sha256': 'x'},
    ], sort_keys=True, separators=(',', ':')).encode()

    header = struct.pack('<4sIII', b'WLRN', 1, len(manifest_json), len(toc_json))
    blob_data = b'\x00' * 5
    bundle = header + manifest_json + toc_json + blob_data

    with pytest.raises(BundleError, match='out of bounds'):
        decode_bundle(bundle)
